python_modules/dagster/dagster/components/core/decl.py

```diff
@@ -0,0 +1,310 @@
+    def __init__(self, context: ComponentLoadContext, path: Path):
+        self.path = path
+
+    def _load_component(self) -> DagsterDefsComponent:
+        return DagsterDefsComponent(path=self.path)
+
+
+class DefsFolderDecl(ComponentDecl[DefsFolderComponent]):
+    def __init__(
+        self, context: ComponentLoadContext, path: Path, children: Mapping[Path, ComponentDecl]
+    ):
+        self.path = path
+        self.children = children
+
+    @classmethod
+    def get(cls, context: ComponentLoadContext) -> "DefsFolderDecl":
+        component = get_component_decl(context)
+        return check.inst(
+            component,
+            DefsFolderDecl,
+            f"Expected PendingDefsFolderComponent at {context.path}, got {component}.",
+        )
+
+    def _load_component(self) -> "DefsFolderComponent":
+        return DefsFolderComponent(
+            path=self.path,
+            children={subpath: decl._load_component() for subpath, decl in self.children.items()},  # noqa: SLF001
+            asset_post_processors=None,
+        )
+
+    def iterate_component_decls(self) -> Iterator[ComponentDecl]:
+        for _, component in self.iterate_path_component_decl_pairs():
+            yield component
+
+    def iterate_path_component_decl_pairs(
+        self,
+    ) -> Iterator[tuple[ComponentPath, ComponentDecl]]:
+        for path, component_node in self.children.items():
+            yield ComponentPath(file_path=path), component_node
+
+            if isinstance(component_node, DefsFolderDecl):
+                yield from component_node.iterate_path_component_decl_pairs()
+
+            if isinstance(component_node, CompositeYamlDecl):
+                for idx, inner_comp in enumerate(component_node.decls):
+                    yield ComponentPath(file_path=path, instance_key=idx), inner_comp
+
+            if isinstance(component_node, CompositePythonDecl):
+                for attr, inner_comp in component_node.decls.items():
+                    yield ComponentPath(file_path=path, instance_key=attr), inner_comp
+
+
+def get_component_decl(context: ComponentLoadContext) -> Optional[ComponentDecl]:
+    """Attempts to determine the type of component that should be loaded for the given context.  Iterates through potential component
+    type matches, prioritizing more specific types: YAML, Python, plain Dagster defs, and component
+    folder.
+    """
+    # in priority order
+    # yaml component
+    if find_defs_or_component_yaml(context.path):
+        return load_yaml_component_decl(context)
+    # pythonic component
+    elif (
+        context.terminate_autoloading_on_keyword_files and (context.path / "component.py").exists()
+    ):
+        return get_component_decl_from_python_file(context)
+    # defs
+    elif (
+        context.terminate_autoloading_on_keyword_files
+        and (context.path / "definitions.py").exists()
+    ):
+        return DagsterDefsDecl(context=context, path=context.path / "definitions.py")
```

Comment:
I want to stop treating this specially. Please add a new component type in follow up and eliminate this clause